File: smart_screens/smart_screens/api/mixing_log.py
import frappe
from frappe import _
from frappe.utils import now

@frappe.whitelist()
def process_batch_scan(scan_lot_number):
    """Process batch scan - get batch details and create mixing log entry"""
    try:
        # First, check if this lot number has already been scanned
        existing_log = frappe.db.exists("Mixing Log", {"scan_lot_number": scan_lot_number})
        
        if existing_log:
            # Get the existing log details to show to user
            existing_data = frappe.get_doc("Mixing Log", existing_log)
            return {
                "success": False,
                "error": f"Lot number '{scan_lot_number}' has already been scanned on {frappe.utils.format_datetime(existing_data.log_datetime)}",
                "duplicate": True,
                "existing_data": {
                    "name": existing_data.name,
                    "scan_lot_number": existing_data.scan_lot_number,
                    "batch_number": existing_data.batch_number,
                    "item_code": existing_data.item_code,
                    "item_name": existing_data.item_name,
                    "item_group": existing_data.item_group,
                    "log_datetime": existing_data.log_datetime
                }
            }
        
        # Get batch details from Stock Entry based on mix_barcode
        batch_details = get_batch_details_from_stock_entry(scan_lot_number)
        
        if not batch_details:
            return {
                "success": False,
                "error": f"No batch found for scanned lot number: {scan_lot_number}"
            }
        
        # Create new Mixing Log document
        mixing_log = frappe.new_doc("Mixing Log")
        mixing_log.scan_lot_number = scan_lot_number
        mixing_log.batch_number = batch_details.get('batch_no')
        mixing_log.item_code = batch_details.get('item_code')
        mixing_log.item_name = batch_details.get('item_name')
        mixing_log.item_group = batch_details.get('item_group')
        mixing_log.log_datetime = now()
        
        # Insert the document
        mixing_log.insert(ignore_permissions=True)
        
        return {
            "success": True,
            "data": {
                "name": mixing_log.name,
                "scan_lot_number": mixing_log.scan_lot_number,
                "batch_number": mixing_log.batch_number,
                "item_code": mixing_log.item_code,
                "item_name": mixing_log.item_name,
                "item_group": mixing_log.item_group,
                "log_datetime": mixing_log.log_datetime
            }
        }
        
    except Exception as e:
        frappe.log_error(f"Error processing batch scan: {str(e)}")
        return {
            "success": False,
            "error": str(e)
        }

def get_batch_details_from_stock_entry(scan_lot_number):
    """Get batch details from Stock Entry based on mix_barcode"""
    try:
        # Search for Stock Entry Detail with matching mix_barcode
        stock_entry_detail = frappe.db.sql("""
            SELECT sed.batch_no, sed.item_code, i.item_name, i.item_group, sed.spp_batch_number
            FROM `tabStock Entry Detail` sed
            LEFT JOIN `tabItem` i ON sed.item_code = i.name
            WHERE sed.mix_barcode = %s
            AND sed.is_finished_item = 1
            ORDER BY sed.creation DESC
            LIMIT 1
        """, (scan_lot_number,), as_dict=True)
        
        if stock_entry_detail:
            return stock_entry_detail[0]
        else:
            return None
            
    except Exception as e:
        frappe.log_error(f"Error in get_batch_details_from_stock_entry: {str(e)}")
        return None

@frappe.whitelist()
def get_recent_logs(limit=10):
    """Get recent mixing log entries"""
    try:
        logs = frappe.get_all(
            "Mixing Log",
            fields=[
                "name",
                "scan_lot_number",
                "batch_number",
                "item_code",
                "item_name",
                "item_group",
                "log_datetime"
            ],
            order_by="log_datetime desc",
            limit=limit
        )
        
        return logs
        
    except Exception as e:
        frappe.log_error(f"Error fetching recent logs: {str(e)}")
        return []

@frappe.whitelist()
def validate_barcode(barcode):
    """Validate if barcode exists in system"""
    try:
        # Check if barcode exists in Stock Entry Details
        exists = frappe.db.sql("""
            SELECT COUNT(*) as count
            FROM `tabStock Entry Detail`
            WHERE mix_barcode = %s
            AND is_finished_item = 1
        """, (barcode,), as_dict=True)
        
        return {
            "success": True,
            "exists": exists[0].get('count', 0) > 0
        }
        
    except Exception as e:
        frappe.log_error(f"Error validating barcode: {str(e)}")
        return {
            "success": False,
            "error": str(e)
        }

File: smart_screens/smart_screens/page/mixing_log/mixing_log.js
// Initialize the page object if it doesn't exist
if (!frappe.pages) {
    frappe.pages = {};
}
if (!frappe.pages['mixing_log']) {
    frappe.pages['mixing_log'] = {};
}

frappe.pages['mixing_log'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Mixing Log',
        single_column: true
    });

    // Create the main content area
    page.main = $('<div class="mixing-log-main"></div>').appendTo(page.body);
    
    // Initialize the mixing log interface
    new MixingLogInterface(page);
};

class MixingLogInterface {
    constructor(page) {
        this.page = page;
        this.setup_interface();
        this.setup_scanner();
    }

    setup_interface() {
        // Create the HTML structure with full width
        this.page.main.html(`
            <div class="mixing-log-container">
                <div class="row">
                    <div class="col-md-6">
                        <div class="card">
                            <div class="card-header">
                                <h4>Mixing Log Scanner</h4>
                            </div>
                            <div class="card-body">
                                <div class="form-group">
                                    <label for="batch-scanner">Batch Scan</label>
                                    <div class="input-group">
                                        <input type="text" class="form-control" id="batch-scanner" 
                                               placeholder="Scan or enter batch barcode" autofocus>
                                        <div class="input-group-btn">
                                            <button class="btn btn-primary" id="manual-entry-btn">
                                                Manual Entry
                                            </button>
                                        </div>
                                    </div>
                                </div>
                                
                                <div class="scan-result" id="scan-result" style="display: none;">
                                    <div class="alert alert-success">
                                        <h5>Scan Result:</h5>
                                        <table class="table table-borderless">
                                            <tr>
                                                <td><strong>Lot Number:</strong></td>
                                                <td id="result-lot-number"></td>
                                            </tr>
                                            <tr>
                                                <td><strong>Batch Number:</strong></td>
                                                <td id="result-batch-number"></td>
                                            </tr>
                                            <tr>
                                                <td><strong>Item Code:</strong></td>
                                                <td id="result-item-code"></td>
                                            </tr>
                                            <tr>
                                                <td><strong>Item Name:</strong></td>
                                                <td id="result-item-name"></td>
                                            </tr>
                                            <tr>
                                                <td><strong>Item Group:</strong></td>
                                                <td id="result-item-group"></td>
                                            </tr>
                                            <tr>
                                                <td><strong>Date & Time:</strong></td>
                                                <td id="result-datetime"></td>
                                            </tr>
                                        </table>
                                    </div>
                                </div>

                                <div class="error-result" id="error-result" style="display: none;">
                                    <div class="alert alert-danger">
                                        <span id="error-message"></span>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                    
                    <!-- Recent Logs in right column -->
                    <div class="col-md-6">
                        <div class="card">
                            <div class="card-header">
                                <h5>Recent Mixing Logs</h5>
                            </div>
                            <div class="card-body">
                                <div id="recent-logs"></div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        `);

        this.load_recent_logs();
    }

    setup_scanner() {
        const scanner_input = this.page.main.find('#batch-scanner');
        const manual_btn = this.page.main.find('#manual-entry-btn');

        // Handle Enter key press for scanning
        scanner_input.on('keypress', (e) => {
            if (e.which === 13) { // Enter key
                const scanned_value = scanner_input.val().trim();
                if (scanned_value) {
                    this.process_scan(scanned_value);
                    scanner_input.val(''); // Clear input after processing
                }
            }
        });

        // Handle manual entry button
        manual_btn.on('click', () => {
            const manual_value = scanner_input.val().trim();
            if (manual_value) {
                this.process_scan(manual_value);
                scanner_input.val(''); // Clear input after processing
            } else {
                frappe.msgprint('Please enter a barcode value');
            }
        });

        // Auto-focus on the input field
        scanner_input.focus();
        
        // Refocus when clicking anywhere on the page
        $(document).on('click', () => {
            setTimeout(() => scanner_input.focus(), 100);
        });
    }

    process_scan(scanned_value) {
        // Show loading state
        this.show_loading();

        // Call API to get batch details and create mixing log
        frappe.call({
            method: 'smart_screens.smart_screens.api.mixing_log.process_batch_scan',
            args: {
                scan_lot_number: scanned_value
            },
            callback: (response) => {
                if (response.message && response.message.success) {
                    this.show_success(response.message.data);
                    this.load_recent_logs(); // Refresh recent logs
                } else if (response.message && response.message.duplicate) {
                    // Handle duplicate scan
                    this.show_error(
                        response.message.error, 
                        true, 
                        response.message.existing_data
                    );
                } else {
                    this.show_error(response.message ? response.message.error : 'Unknown error occurred');
                }
            },
            error: (error) => {
                this.show_error(error.message || 'Failed to process scan');
            }
        });
    }

    show_loading() {
        this.page.main.find('#scan-result').hide();
        this.page.main.find('#error-result').hide();
        frappe.show_progress('Processing scan...', 50, 100);
    }

    show_success(data) {
        frappe.hide_progress();
        
        // Populate result fields
        this.page.main.find('#result-lot-number').text(data.scan_lot_number);
        this.page.main.find('#result-batch-number').text(data.batch_number);
        this.page.main.find('#result-item-code').text(data.item_code);
        this.page.main.find('#result-item-name').text(data.item_name);
        this.page.main.find('#result-item-group').text(data.item_group);
        this.page.main.find('#result-datetime').text(frappe.datetime.str_to_user(data.log_datetime));

        // Show success result
        this.page.main.find('#error-result').hide();
        this.page.main.find('#scan-result').show();

        // Auto-hide after 3 seconds and refocus
        setTimeout(() => {
            this.page.main.find('#scan-result').fadeOut();
            this.page.main.find('#batch-scanner').focus();
        }, 3000);
    }

    show_error(error_message, is_duplicate = false, existing_data = null) {
        frappe.hide_progress();
        
        if (is_duplicate && existing_data) {
            // Show duplicate warning with existing data
            this.page.main.find('#error-result').removeClass('alert alert-danger').addClass('alert alert-warning');
            
            let duplicate_html = `
                <strong>Duplicate Scan Detected!</strong><br>
                <span>${error_message}</span><br><br>
                <strong>Previously scanned details:</strong><br>
                <table class="table table-borderless" style="margin-top: 10px;">
                    <tr><td><strong>Lot Number:</strong></td><td>${existing_data.scan_lot_number}</td></tr>
                    <tr><td><strong>Batch Number:</strong></td><td>${existing_data.batch_number}</td></tr>
                    <tr><td><strong>Item Code:</strong></td><td>${existing_data.item_code}</td></tr>
                    <tr><td><strong>Item Name:</strong></td><td>${existing_data.item_name}</td></tr>
                    <tr><td><strong>Item Group:</strong></td><td>${existing_data.item_group}</td></tr>
                    <tr><td><strong>Original Scan Time:</strong></td><td>${frappe.datetime.str_to_user(existing_data.log_datetime)}</td></tr>
                </table>
            `;
            this.page.main.find('#error-message').html(duplicate_html);
        } else {
            // Show regular error
            this.page.main.find('#error-result').removeClass('alert alert-warning').addClass('alert alert-danger');
            this.page.main.find('#error-message').text(error_message);
        }
        
        this.page.main.find('#scan-result').hide();
        this.page.main.find('#error-result').show();

        // Auto-hide error after 8 seconds for duplicates (longer to read), 5 seconds for regular errors
        const hideTimeout = is_duplicate ? 8000 : 5000;
        setTimeout(() => {
            this.page.main.find('#error-result').fadeOut();
            this.page.main.find('#batch-scanner').focus();
        }, hideTimeout);
    }

    load_recent_logs() {
        frappe.call({
            method: 'smart_screens.smart_screens.api.mixing_log.get_recent_logs',
            callback: (response) => {
                if (response.message) {
                    this.render_recent_logs(response.message);
                }
            }
        });
    }

    render_recent_logs(logs) {
        const logs_container = this.page.main.find('#recent-logs');
        
        if (logs.length === 0) {
            logs_container.html('<p class="text-muted">No recent logs found.</p>');
            return;
        }

        let html = `
            <div class="table-responsive">
                <table class="table table-striped table-condensed">
                    <thead>
                        <tr>
                            <th>Lot Number</th>
                            <th>Batch Number</th>
                            <th>Item Code</th>
                            <th>Item Group</th>
                            <th>Date & Time</th>
                        </tr>
                    </thead>
                    <tbody>
        `;

        logs.forEach(log => {
            html += `
                <tr>
                    <td>${log.scan_lot_number}</td>
                    <td>${log.batch_number || ''}</td>
                    <td>${log.item_code || ''}</td>
                    <td>${log.item_group || ''}</td>
                    <td>${frappe.datetime.str_to_user(log.log_datetime)}</td>
                </tr>
            `;
        });

        html += '</tbody></table></div>';
        logs_container.html(html);
    }
}